### ISSUE TITLE
Using time as hyperbrand budget

### ISSUE THREAD

**Issue by ggenesum** [optuna/optuna#2115]
Hi everyone,

When using   

```
study = optuna.create_study(study_name='PPO2-study', direction='maximize',
    pruner=optuna.pruners.MedianPruner(
        n_startup_trials=4, n_warmup_steps=2, interval_steps=1))
```

then in objective function

```

    def objective(trial):
          #set model hyperparameters from trial object
          # ...
          for i in range(7):      
             model.learn(total_timesteps=200000)
             reward = model.test()
             print("intermediate reward:" + str(reward))
             trial.report(reward, i)
             if trial.should_prune():
                raise optuna.TrialPruned()
```

Trial can be pruned from 200000 iterations, is something wrong with my code? I want at least 400000 timesteps before pruning the trial.

Thanks and happy tuning

**Comment by HideakiImamura** [comment]
Hmm... At the first glance, `optuna.pruners.MedianPruner`'s pruning check should start at step 2 (after finishing 400000 timesteps as you wanted). I think it may depend on your `model.learn()` call.

**Comment by ggenesum** [comment]
The median pruner example with sklearn is working normaly,
I think optuna is doing fine and something is wrong with timestep logging on tensorboard, I had issues with it because of multithreading. 

Anyway, I decided to use HyperBrand with elapsed training time as budget  because training speed differ a lot depending of hyperparameters. Is there anythink wrong with this?

**Comment by HideakiImamura** [comment]
> I think optuna is doing fine and something is wrong with timestep logging on tensorboard, I had issues with it because of multithreading.

OK. You are welcome to create a new issue with details of the problem and reproducible code examples.

> Anyway, I decided to use HyperBrand with elapsed training time as budget because training speed differ a lot depending of hyperparameters. Is there anything wrong with this?

It is a good idea to use HyperBand as a budget for the elapsed time. However, you have to be careful: Optuna's pruner with HyperBand calculates the budget as a unit of `trial.report(value, step)` steps. Therefore, if you want to use the elapsed time as a budget, you need to define a unit of budget (e.g. 1sec), calculate the step, and do `trial.report(value, step)`. The following is an example code.
```python
def objective(trial):
    # sample hyperparameters
    param = trial.suggest_float("name", 0., 1.)
    
    # train your model with reporting and pruning
    start = time.time()
    for i in range(n_iterations):
        value = partially_evaluate_your_model(...)
        accumulated_elapsed_time = time.time() - start
        
        trial.report(value, int(accumulated_elapsed_time))
        if trial.should_prune():
            raise optuna.TrialPruned()

    return value
```
More ideally, the function that trains the model should be able to pass a callback function that computes intermediate values and `trial.report(value, step)` at constant intervals (unit time). For example, the following is an example of the code on the Optuna side.
```python
def objective(trial):
    # sample hyperparameters
    param = trial.suggest_float("name", 0., 1.)
    
    # define the callback function
    # this function is called inside the model training, that is, `partially_evaluate_your_model` 
    def callback(accumulated_elapsed_time, ...):
        trial.report(value, int(accumulated_elapsed_time))
        if trial.should_prune():
            raise optuna.TrialPruned()

    for i in range(n_iterations):
        # Assume that `partially_evaluate_your_model` measures time in its execution 
        # and calls a callback function every constant time.
        value = partially_evaluate_your_model(..., callback=callback)

    return value
```

**Comment by ggenesum** [comment]
Thank you.
Actually tensorboard logging is completly unrelated to optuna.

I used something like this to implement time as budget with no callback

`
def objective(trial):
    total_time_elapsed = 0
    while (total_time_elapsed < 10000):
        start_time = time.clock()
        model.learn(total_timesteps=50000, reset_num_timesteps = False)
        total_time_elapsed += time.clock() - start_time
        if (total_time_elapsed > 10000): #clip last intermediate report timestep
            total_time_elapsed = 10000
     
        reward = model.test()
        trial.report(reward, total_time_elapsed)
        if trial.should_prune():
            raise optuna.TrialPruned()

    return reward
`

I'm not sure if clipping the last timestep is necessary. It seems to be working normaly without it.

From what I know about the library I'm using to create the model (Stable-Baselines), it will be difficult to synchronise report steps in time. 
Again, it seems pruning is working normaly even with different report steps upon different models, but the problem I want to solve is quite easy, I guess even random search would work so I'm not totally sure about pruning.
